### PR TITLE
Animation - Player States - Factories - Input

### PR DESCRIPTION
**Pull request by Abar23**
Animation Rework - #11

### CODE REVIEW

File: src/core/ecs/systems/PlayerSystem.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+			
+			//updates the animation
+			if (newType != PlayerType::Unchanged) {
+				PlayerType newDirectedType = PlayerFactory::getDirectionalType(newType, player.input->getCurrentDirection());
+				PlayerFactory::setAnimationData(newDirectedType, animation, animationPlayer);
+			}
+
+			if (animation.animationType == AnimationType::Manual) {
+				if (player.playerState->getType() == Sliding) {
+					if (player.playerState->getSlideState() == PlayerState::StandingUp) {
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/core/ecs/systems/PlayerSystem.h:46]: getSlideState is virtual, the sliding state implementation can return StandingUp

File: src/core/ecs/systems/PlayerSystem.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+						animation.currentFrameIndex = 1;
+					}
+				}
+				else {
+					std::cout << "Unhandled manual animation" << std::endl;
+				}
+			}
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/core/ecs/systems/PlayerSystem.h:53]: Some anims are coupled with game logic so they can't use delta time to determine things. If something is declared as manual (needing manual frame updates) but isn't handled here, then something is wrong

File: src/game_logic/factories/Factory.h [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+#include "../../core/ecs/components/Animation.h"
+#include "../../core/ecs/components/AnimationPlayer.h"
+
+class Factory {
+
+public:
+	virtual Animation getAnimationData(enum type, Animation oldAnim) = 0;
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/game_logic/factories/Factory.h]: fixed it

File: src/game_logic/factories/PlayerFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+	animationP.frameOffset = 0;
+	animationP.frameList = { 132 };
+
+	switch (type)
+	{
+	case Standing:
+		animation.animationType = AnimationType::Static;
+
+		animation.currentFrameIndex = 0;
+		animation.frameTime = 9999999;
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/game_logic/factories/PlayerFactory.cpp]: sure, done

File: src/game_logic/factories/PlayerFactory.cpp [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+	}
+
+	// reset animation player data
+	animationP.animationType = AnimationType::Static;
+
+	animationP.currentThrowFrameIndex = 0;
+	animationP.throwFrameTime = 1;
+	animationP.throwFrameTimeCounter = 0;
+	animationP.frameOffset = 0;
+	animationP.frameList = { 132 };
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/game_logic/factories/PlayerFactory.cpp:86]: it needs to be there, I can explain it irl sometime if you want

File: src/core/ecs/components/SpriteRenderer.h [hunk 6]
```diff
@@ -23,5 +23,7 @@ struct SpriteRenderer
 	float frameWidth;
 	float frameHeight;
 
-	vec3 cameraPosition; //this is temporary
+	bool flipX;
+
+	vec3 offset = vec3();
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/core/ecs/components/SpriteRenderer.h]: sure

File: src/core/ecs/systems/AnimationSystem.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+#include "core/ecs/components/Player.h"
+
+class AnimationSystem : public AbstractSystem
+{
+private:
+
+
+
+public:
+	void Init() {
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/core/ecs/systems/AnimationSystem.h]: nope

File: src/core/ecs/systems/AnimationSystem.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+				}
+				else {
+					anim.currentFrameIndex = 0;
+					anim.flipX = true;
+
+					std::vector<int> newFrameList = std::vector<int>();
+					for (int i = anim.frameList.size()-2; i > 0; --i) {
+						newFrameList.push_back(anim.frameList[i]);
+					}
+					anim.frameList = newFrameList;
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/core/ecs/systems/AnimationSystem.h]: understood
new vectors = bad

File: src/game_logic/states/player_states/PlayerState.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
+#pragma once
+#include <vector>
+#include <iostream>
+#include "Input.h"
+#include "math/Vec3.h"
+
+#include "game_logic/factories/PlayerTypes.h"
+
+class PlayerState {
```
[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/game_logic/states/player_states/PlayerState.h:9]: Easier said than done, its a bit of a knot so it'll take a lil bit

[the remark this answers is not in the document]
Comment (ConsiderTheSun) [reply src/game_logic/states/player_states/PlayerState.h:9]: Actually shouldn't be toooo hard, I just need to pass in the entity like we talked about
should we merge in your changes first tho?

[the remark this answers is not in the document]
Comment (Abar23) [reply src/game_logic/states/player_states/PlayerState.h:9]: Sure. Make sure you have any changes currently on your machine pushed up. I'll pull those into my branch and we can review the PR.

**Pull request merged**